### PR TITLE
ROX-20749: RHTAP - central-db onboarding

### PR DESCRIPTION
**Pull request by tommartensen**
## Description

Onboarding central-db image build to RHTAP. 

## Checklist
- [ ] Investigated and inspected CI test results
- ~[ ] Unit test and regression tests added~
- ~[ ] Evaluated and added CHANGELOG entry if required~
- ~[ ] Determined and documented upgrade steps~
- ~[ ] Documented user facing changes (create PR based on [openshift/openshift-docs](https://github.com/openshift/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))~

If any of these don't apply, please comment below.

## Testing Performed

Passing RHTAP pipeline + Enterprise contract.

### CODE REVIEW

File: .tekton/central-db-pull-request.yaml
Comment (tommartensen) [inline .tekton/central-db-pull-request.yaml:1]: The newer `apiVersion`, `on-cel-expression` instead of the trigger was preset by RHTAP's generated PR.

File: .tekton/central-db-pull-request.yaml
Comment (tommartensen) [inline .tekton/central-db-pull-request.yaml:272]: This is a new task from the RHTAP generated pipeline, which we may need to enable and roll out to other pipelines, if EC requires it.

File: image/postgres/rhtap.Dockerfile
Comment (tommartensen) [inline image/postgres/rhtap.Dockerfile:1]: Based on https://gitlab.cee.redhat.com/stackrox/rhacs-midstream/-/blob/rhacs-1.0-rhel-8/distgit/containers/rhacs-central-db/Dockerfile.in

File: image/postgres/rhtap.Dockerfile
Comment (tommartensen) [inline image/postgres/rhtap.Dockerfile]: Is there an authoritative source for these descriptions, and tags? 
This information seems to be propagated to the catalog, e.g. for 4.2.2: https://catalog.redhat.com/software/containers/advanced-cluster-security/rhacs-central-db-rhel8/635face27140d313bbc4afbd?architecture=amd64&image=653112d59841e831b5558a38&container-tabs=technical-information .

File: image/postgres/rhtap.Dockerfile
Comment (msugakov) [reply image/postgres/rhtap.Dockerfile:1]: Up to you, of course, but you can avoid copying the entire source code and having this `builder` stage altogether by adjusting `COPY` commands in the `final` stage: take files from the context instead of the `builder` stage.

The `builder` exists in downstream dockerfile because it's most canonical to use provided environment variables (`$REMOTE_SOURCES` and `$REMOTE_SOURCES_DIR`) that way. Since these variables don't exist and aren't needed in RHTAP, we can do it simpler here.

File: image/postgres/rhtap.Dockerfile
Comment (msugakov) [inline image/postgres/rhtap.Dockerfile:5]: Let's make it explicit that we use the `latest` tag.

```suggestion
FROM registry.redhat.io/rhel8/postgresql-13:latest AS final
```

File: .tekton/central-db-pull-request.yaml
Comment (msugakov) [inline .tekton/central-db-pull-request.yaml:11]: Optional suggestion.

```suggestion
    pipelinesascode.tekton.dev/on-event: '[pull_request]'
    pipelinesascode.tekton.dev/on-target-branch: '[*]'
```

This should give the same effect but will be uniform with `roxctl` pipeline.
Either way we decide, let's please keep all pipelines similar.

File: .tekton/central-db-pull-request.yaml
Comment (msugakov) [inline .tekton/central-db-pull-request.yaml:40]: WDYT about adding empty `prefetch-input` for explicitness?

```suggestion
  # No language dependencies are required for central-db image.
  - name: prefetch-input
    value: ''

```

(similarly in the push pipeline)

File: .tekton/central-db-pull-request.yaml
Comment (msugakov) [inline .tekton/central-db-pull-request.yaml:142]: This looks like a new thing, not present in `roxctl` pipeline. RHTAP don't yet offer ways to keep pipelines up-to-date so we need to do that manually for now.

How do we best approach this: as a one-off after onboarding all other containers or on an ongoing basis?

File: .tekton/central-db-pull-request.yaml
Comment (msugakov) [inline .tekton/central-db-pull-request.yaml:192]:
```suggestion
        value: $(params.revision)
      # A shallow repo clone is sufficient for central-db build.
      - name: depth
        value: "1"
      - name: fetchTags
        value: "false"
```

(similarly in the push pipeline)

File: .tekton/central-db-pull-request.yaml
Comment (msugakov) [inline .tekton/central-db-pull-request.yaml:232]: This isn't needed due to how prefetch task is written.
```suggestion
```

(similarly in the push pipeline)

File: .tekton/central-db-pull-request.yaml
Comment (msugakov) [reply .tekton/central-db-pull-request.yaml:272]: Even if EC does not require it, we must provide source images for compliance with the Red Hat Open Source policies. If this step gives us that, we should add it to other pipelines eventually.

File: .tekton/central-db-pull-request.yaml
Comment (tommartensen) [reply .tekton/central-db-pull-request.yaml:142]: Discussed in weekly meeting to have a reconcile at the end of all onboardings: https://issues.redhat.com/browse/ROX-20814. 
I would postpone it to that task.

File: .tekton/central-db-pull-request.yaml
Comment (tommartensen) [reply .tekton/central-db-pull-request.yaml:11]: Discussed in weekly meeting to have a reconcile at the end of all onboardings: https://issues.redhat.com/browse/ROX-20814. 
I would postpone it to that task.

File: .tekton/central-db-pull-request.yaml
Comment (tommartensen) [reply .tekton/central-db-pull-request.yaml:43]: added for expliciticy

File: .tekton/central-db-pull-request.yaml
Comment (tommartensen) [reply .tekton/central-db-pull-request.yaml:195]: added for explicitness

File: .tekton/central-db-pull-request.yaml
Comment (tommartensen) [reply .tekton/central-db-pull-request.yaml:276]: Discussed in weekly meeting to have a reconcile at the end of all onboardings: https://issues.redhat.com/browse/ROX-20814.
I would postpone it to that task.

File: .tekton/central-db-pull-request.yaml
Comment (tommartensen) [reply .tekton/central-db-pull-request.yaml]: removed

File: image/postgres/rhtap.Dockerfile
Comment (msugakov) [inline image/postgres/rhtap.Dockerfile:39]: I hate doing this, but I'd ask you to check with the Core Workflows team, probably @connorgorman and @c-du, why at all do we do `--chown=postgres:postgres` on `docker-entrypoint.sh` and don't do that on `init-entrypoint.sh` in the [downstream dockerfile](https://gitlab.cee.redhat.com/stackrox/rhacs-midstream/-/blob/rhacs-1.0-rhel-8/distgit/containers/rhacs-central-db/Dockerfile.in?ref_type=heads#L41).

Even though it's a negligible risk, having scripts mutable by the user from whom the container executes gives a way to inject custom code in the container.

Note that after you moved `USER 70:70` above, `init-entrypoint.sh` will be owned by `postgres:postgres` unlike how it was owned by `root:root` before.

File: image/postgres/rhtap.Dockerfile
Comment (msugakov) [reply image/postgres/rhtap.Dockerfile]: I did not know about that feature. I don't know the answer. Would suggest to ask in `pyxis` GChat.

File: image/postgres/rhtap.Dockerfile
Comment (tommartensen) [reply image/postgres/rhtap.Dockerfile:39]: Yes, I was aware of this change. Asked in Slack: https://redhat-internal.slack.com/archives/C01DQH7UP8C/p1699948813746169.

File: image/postgres/rhtap.Dockerfile
Comment (tommartensen) [reply image/postgres/rhtap.Dockerfile]: The required labels are explained [here](https://enterprisecontract.dev/docs/rhtap-policy/index.html#_required_oci_image_labels). 
Their values match what we have defined on midstream today. 
I created another reconcile ticket for this: https://issues.redhat.com/browse/ROX-20839.

File: image/postgres/rhtap.Dockerfile
Comment (tommartensen) [reply image/postgres/rhtap.Dockerfile:37]: TLDR of Slack conversation:

```
# docker run -it --entrypoint /bin/sh quay.io/redhat-user-workloads/rh-acs-tenant/acs/central-db:on-pr-de7e68e6a4e98b892fff8ce8148619d5ae32c67c
sh-4.4$ ls -lisa /usr/local/bin/
total 28
492156  4 drwxr-xr-x 1 root root  4096 Nov 15 09:44 .
492155  4 drwxr-xr-x 1 root root  4096 Oct 23 10:35 ..
491489 16 -rwxrwxr-x 1 root root 12349 Nov 15 09:33 docker-entrypoint.sh
491490  4 -rwxrwxr-x 1 root root   249 Nov 15 09:33 init-entrypoint.sh
sh-4.4$ docker-entrypoint.sh
Waiting for child process 16 to exit
flock: bad file descriptor: '/var/lib/postgresql/data/pglock'
sh-4.4$ init-entrypoint.sh
initdb: error: no data directory specified
You must identify the directory where the data for this database system
will reside.  Do this with either the invocation option -D or the
environment variable PGDATA.
sh-4.4$ whoami
postgres
sh-4.4$ id
uid=70(postgres) gid=70(postgres) groups=70(postgres)
```

This was my procedure to ensure the scripts can be executed by `postgres`, when owned by `root`. Also, the scripts are not world-writeable (`-rwxrwxr-x`), so if a malicious user came in as the postgres user, they won’t be able to modify the scripts.

File: image/postgres/rhtap.Dockerfile
Comment (tommartensen) [reply image/postgres/rhtap.Dockerfile]: Simplified to one stage.

File: image/postgres/rhtap.Dockerfile
Comment (msugakov) [reply image/postgres/rhtap.Dockerfile]: Required labels for Enterprise Contract may overlap with labels used by Red Hat Catalog, but I highly doubt these are the same sets (or that the catalog ones are a subset of EC labels).

File: .tekton/central-db-pull-request.yaml
Comment (msugakov) [inline .tekton/central-db-pull-request.yaml:145]: After glancing at the pipeline I realized that `build-source-image` task does not run.

image

Do you mind setting `build-source-image` to `"true"` in `spec.params`?

If we get this step running here without much troubleshooting, let's do it right away and have less follow-ups to do later.

**Review by msugakov — COMMENTED** [review]
I was about to approve when got curious to check something in the build logs and noticed that source build is skipped.

File: .tekton/central-db-pull-request.yaml
Comment (tommartensen) [reply .tekton/central-db-pull-request.yaml]: The source image is created a warning: `Image registry.redhat.io/rhel8/postgresql-13:latest does not come from supported registry. Skip handling the sources from it.`

Result

```
{
  "status": "success", 
  "dependencies_included": false, 
  "base_image_source_included": false, 
  "image_url": "quay.io/redhat-user-workloads/rh-acs-tenant/acs/central-db:on-pr-4b8e0ed074adac9f26be380eb794d1f7ba46b10d.src", 
  "image_digest": "sha256:4c63cdc5e5e035250fcb1d0f703ad792cc5ba483bba8e67e9e7a6e2936a845fb"
}
```

File: .tekton/central-db-pull-request.yaml
Comment (msugakov) [inline .tekton/central-db-pull-request.yaml:142]: Sorry, but I meant not changing the default here but adding a parameter in `spec.params`, lines 22-42. Doing that way should let us more easily see what we customized.

Comment (tommartensen) [reply .tekton/central-db-pull-request.yaml:142]: Sure, but what do you think about the result of the test?

Comment (msugakov) [reply .tekton/central-db-pull-request.yaml:142]: You mean that warning `Image registry.redhat.io/rhel8/postgresql-13:latest does not come from supported registry. Skip handling the sources from it.`? I overlooked it in logs since it has `INFO` on it, noticed only now :-(

I've seen a PR auto-opened with task updates. Not sure if sources task is there. I'd suggest to check if an updated task fixes that. If not, please report it to Slack, but I wouldn't block the PR on it because it's not failing the step.

Comment (tommartensen) [reply .tekton/central-db-pull-request.yaml:142]: `registry.access.redhat.com` is the only supported registry for the base image: https://github.com/redhat-appstudio/build-definitions/blob/main/containers/source-build/app/source_build.py#L185-L190. 
I'll ask on Slack.

**Review by msugakov — APPROVED** [review]

**Pull request merged**
